test/token_holder/redeem.js

```diff
@@ -0,0 +1,833 @@
+const UtilityTokenMock = artifacts.require('UtilityTokenMock');
+
+const ephemeralKeyAddress1 = '0x62502C4DF73935D0D10054b0Fb8cC036534C6fb0';
+const ephemeralPrivateKey1 = '0xa8225c01ceeaf01d7bc7c1b1b929037bd4050967c5730c0b854263121b8399f3';
+const ephemeralPrivateKey2 = '0x634011a05b2f48e2d19aba49a9dbc12766bf7dbd6111ed2abb2621c92e8cfad9';
+
+let token;
+
+
+function getRedeemCallPrefix() {
+  return web3.eth.abi.encodeFunctionSignature({
+    name: 'redeem',
+    type: 'function',
+    inputs: [
+      {
+        type: 'uint256', name: '',
+      },
+      {
+        type: 'address', name: '',
+      },
+      {
+        type: 'uint256', name: '',
+      },
+      {
+        type: 'uint256', name: '',
+      },
+      {
+        type: 'uint256', name: '',
+      },
+      {
+        type: 'bytes32', name: '',
+      },
+      {
+        type: 'uint256', name: '',
+      },
+      {
+        type: 'uint8', name: '',
+      },
+      {
+        type: 'bytes32', name: '',
+      },
+      {
+        type: 'bytes32', name: '',
+      },
+    ],
+  });
+}
+
+function getRedeemSignedData(
+  tokenHolderAddress, ruleAddress, redeemCallData, nonce, ephemeralKey
+) {
+  const msgHash = web3.utils.soliditySha3(
+    {
+      t: 'bytes1', v: '0x19',
+    },
+    {
+      t: 'bytes1', v: '0x0',
+    },
+    {
+      t: 'address', v: tokenHolderAddress,
+    },
+    {
+      t: 'address', v: ruleAddress,
+    },
+    {
+      t: 'uint8', v: 0,
+    },
+    {
+      t: 'bytes32', v: web3.utils.keccak256(redeemCallData),
+    },
+    {
+      t: 'uint256', v: nonce,
+    },
+    {
+      t: 'uint8', v: 0,
+    },
+    {
+      t: 'uint8', v: 0,
+    },
+    {
+      t: 'uint8', v: 0,
+    },
+    {
+      t: 'bytes4', v: getRedeemCallPrefix(),
+    },
+    {
+      t: 'uint8', v: 0,
+    },
+    {
+      t: 'bytes32', v: '0x0',
+    },
+  );
+
+  const rsv = EthUtils.ecsign(
+    EthUtils.toBuffer(msgHash),
+    EthUtils.toBuffer(ephemeralKey),
+  );
+
+  return { msgHash, rsv };
+}
+
+function getRedeemDataToSign(
+  amount, beneficiary, gasPrice, gasLimit, redeemerNonce) {
+  return web3.eth.abi.encodeFunctionCall(
+    {
+      name: 'redeem',
+      type: 'function',
+      inputs: [
+        {
+          type: 'uint256',
+          name: 'amount',
+        },
+        {
+          type: 'address',
+          name: 'beneficiary',
+        },
+        {
+          type: 'uint256',
+          name: 'gasPrice',
+        },
+        {
+          type: 'uint256',
+          name: 'gasLimit',
+        },
+        {
+          type: 'uint256',
+          name: 'redeemerNonce',
+        },
+      ],
+    },
+    [amount, beneficiary, gasPrice, gasLimit, redeemerNonce],
+  );
+}
+
+async function prepareRedeemPayableRule(
+    accountProvider,
+    amount,
+    beneficiary,
+    facilitator,
+    gasPrice,
+    gasLimit,
+    redeemerNonce,
+    tokenHolder,
+    nonce,
+    ephemeralKey,
+)
+{
+  const mockRule = await MockRule.new();
```

Comment:
We have "MockRule" contract which we use for mocking rules. MockRule contract also being used in multiple tests. e.g. executeRule: https://github.com/OpenSTFoundation/openst-contracts/blob/develop/test/token_holder/execute_rule.js#L23

Here coGateway is acting as rule which is being mocked in MockRule. Variable name correctly represents contract name. 
I also thought of mocking it in separate CoGatewayMock contract but find it maintenance wise, it's too much effort. 

Do you still think it should be renamed?

Reply:
I would say that given what executeRule is meant to do (pass arbitrary data, that may or may not be for a function call, to an arbitrary address) and what redeem is effectively meant to do (call a specific function at a specific address), it would be reasonable for these mocks to be different files.

However, specifically here, all I meant is that "mockRule" (lower case, name of the constant) could be named "coGateway", regardless of the fact that it is an instance of the "MockRule" contract; I think this change would aid comprehension of the test.

My apologies, I was not as clear in this comment as I could have been.